onnxruntime/core/providers/cpu/ml/linearregressor.cc

```diff
@@ -1,61 +1,48 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
-
 #include "core/providers/cpu/ml/linearregressor.h"
+#include "core/util/eigen_common_wrapper.h"
 
 namespace onnxruntime {
 namespace ml {
 
-ONNX_CPU_OPERATOR_ML_KERNEL(
-    LinearRegressor,
-    1,
-    KernelDefBuilder().TypeConstraint("T", DataTypeImpl::GetTensorType<float>()),
-    LinearRegressor<float>);
+ONNX_OPERATOR_TYPED_KERNEL_EX(LinearRegressor, kMLDomain, 1, float, kCpuExecutionProvider,
+                              KernelDefBuilder().TypeConstraint("T", DataTypeImpl::GetTensorType<float>()),
+                              LinearRegressor<float>);
+
+ONNX_OPERATOR_TYPED_KERNEL_EX(LinearRegressor, kMLDomain, 1, double, kCpuExecutionProvider,
+                              KernelDefBuilder().TypeConstraint("T", DataTypeImpl::GetTensorType<double>()),
+                              LinearRegressor<double>);
 
 template <typename T>
-LinearRegressor<T>::LinearRegressor(const OpKernelInfo& info) : OpKernel(info),
-                                                                intercepts_(info.GetAttrsOrDefault<float>("intercepts")),
-                                                                post_transform_(MakeTransform(info.GetAttrOrDefault<std::string>("post_transform", "NONE"))) {
+LinearRegressor<T>::LinearRegressor(const OpKernelInfo& info)
+    : OpKernel(info), post_transform_(MakeTransform(info.GetAttrOrDefault<std::string>("post_transform", "NONE"))) {
+  std::vector<float> c;
+  c = info.GetAttrsOrDefault<float>("intercepts");
+  intercepts_.resize(c.size());
+  std::copy_n(c.data(), c.size(), intercepts_.data());
   ORT_ENFORCE(info.GetAttr<int64_t>("targets", &targets_).IsOK());
-  ORT_ENFORCE(info.GetAttrs<float>("coefficients", coefficients_).IsOK());
-}
-
-template <>
-Status LinearRegressor<float>::Compute(OpKernelContext* ctx) const {
-  const auto* X = ctx->Input<Tensor>(0);
-  if (X->Shape().NumDimensions() == 0) {
-    return Status(common::ONNXRUNTIME, common::INVALID_ARGUMENT,
-                  "Input shape needs to be at least a single dimension.");
+  c.clear();
+  ORT_ENFORCE(info.GetAttrs<float>("coefficients", c).IsOK());
+  coefficients_.resize(c.size());
+  ORT_ENFORCE(c.size() % targets_ == 0);
+  int64_t feature_size = static_cast<int64_t>(c.size()) / targets_;
+  for (int64_t i = 0; i != targets_; ++i) {
+    int64_t offset = i * feature_size;
+    for (int64_t j = 0; j != feature_size; ++j) {
```

Comment:
Maybe the use of pointers would help here, at least for c, for coefficients_ too if targets_ is 1.

Reply:
Thanks, will try.

Reply:
Hi , I wrote a small test, and it didn't show any perf difference.  Pointer is not faster.

```C++
static void BM_Transpose1(benchmark::State& state) {
  int targets_ = 1024;
  std::vector<float> c(1024 * 10);
  std::vector<float> coefficients_(1024 * 10);
  std::uniform_real_distribution<float> d(-1, 1);
  for (size_t i = 0; i != coefficients_.size(); ++i) {
    c[i] = d(mt);
  }
  ORT_ENFORCE(c.size() % targets_ == 0);
  for (auto _ : state) {
    int64_t feature_size = static_cast<int64_t>(c.size()) / targets_;
    for (int64_t i = 0; i != targets_; ++i) {
      int64_t offset = i * feature_size;
      for (int64_t j = 0; j != feature_size; ++j) {
        coefficients_[j * targets_ + i] = c[offset + j];
      }
    }
  }
}

static void BM_Transpose2(benchmark::State& state) {
  int targets_ = 1024;
  std::vector<float> c(1024 * 10);
  std::vector<float> coefficients_(1024 * 10);
  std::uniform_real_distribution<float> d(-1, 1);
  for (size_t i = 0; i != coefficients_.size(); ++i) {
    c[i] = d(mt);
  }
  ORT_ENFORCE(c.size() % targets_ == 0);
  for (auto _ : state) {
    int64_t feature_size = static_cast<int64_t>(c.size()) / targets_;
    for (int64_t i = 0; i != targets_; ++i) {
      float* p1 = coefficients_.data() + i;
      float* p2 = c.data() + i * feature_size;
      for (int64_t j = 0; j != feature_size; ++j) {
        p1[j * targets_] = p2[j];
      }
    }
  }
}
BENCHMARK(BM_Transpose1);
BENCHMARK(BM_Transpose2);
```